tests/test_pytorch/test_fsdp.py

```diff
@@ -0,0 +1,791 @@
+
+    if trainer.global_rank != 0:
+        assert len(full_state_dict) == 0
+        return
+
+    # State dict should contain same number of keys
+    assert len(correct_state_dict) == len(full_state_dict)
+    # OrderedDict should return the same keys in the same order
+    assert all(_ex == _co for _ex, _co in zip(full_state_dict.keys(), correct_state_dict.keys()))
+
+
+def test_fsdp_strategy_cpu_offload():
+    """Test the different ways cpu offloading can be enabled."""
+    # bool
+    strategy = HPUFSDPStrategy(cpu_offload=True)
+    assert strategy.cpu_offload == CPUOffload(offload_params=True)
+
+    # dataclass
+    config = CPUOffload()
+    strategy = HPUFSDPStrategy(cpu_offload=config)
+    assert strategy.cpu_offload == config
+
+
+@pytest.mark.parametrize(
+    ("precision", "expected_dtype"),
+    [
+        ("32-true", torch.float32),
+    ],
+)
+@pytest.mark.xfail(run=False, reason="TODO")
+def test_configure_model(tmpdir, hpus, precision, expected_dtype):
+    """Test that the module under configure_model gets moved to the right device and dtype."""
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        accelerator=HPUAccelerator(),
+        devices=hpus,
+        strategy=HPUFSDPStrategy(
+            parallel_devices=[torch.device("hpu")] * hpus,
+            sharding_strategy="SHARD_GRAD_OP",
+            precision_plugin=HPUFSDPPrecision(precision),
+        ),
+        max_epochs=1,
+        enable_checkpointing=False,
+        sync_batchnorm=True,
+    )
+
+    class MyModel(BoringModel):
+        def configure_model(self):
+            self.layer = torch.nn.Linear(32, 2)
+            expected_device = torch.device("cpu")
+            assert self.layer.weight.device == expected_device
+            assert self.layer.weight.dtype == expected_dtype
+
+        def configure_optimizers(self):
+            return torch.optim.AdamW(self.layer.parameters(), lr=0.1)
+
+        def on_fit_start(self):
+            assert self.layer.weight.device == torch.device("hpu", self.local_rank)
+            assert self.layer.weight.dtype == expected_dtype
+
+    model = MyModel()
+    trainer.fit(model)
+
+
+def test_fsdp_sharding_strategy():
+    """Test the different ways the sharding strategy can be set."""
+    from torch.distributed.fsdp import ShardingStrategy
+
+    # default
+    strategy = HPUFSDPStrategy()
+    assert strategy.sharding_strategy == ShardingStrategy.FULL_SHARD
+
+    # enum
+    strategy = HPUFSDPStrategy(sharding_strategy=ShardingStrategy.SHARD_GRAD_OP)
+    assert strategy.sharding_strategy == ShardingStrategy.SHARD_GRAD_OP
+
+    # string
+    strategy = HPUFSDPStrategy(sharding_strategy="NO_SHARD")
+    assert strategy.sharding_strategy == ShardingStrategy.NO_SHARD
+    strategy = HPUFSDPStrategy(sharding_strategy="no_shard")
+    assert strategy.sharding_strategy == ShardingStrategy.NO_SHARD
+
+
+def test_fsdp_activation_checkpointing():
```

Comment:
this CP works ?

Reply:
This test passes but its only doing API level validation with mock patch and with out any training loop.